### DISCUSSION TITLE
How are Server Actions exposed ?

### DISCUSSION THREAD

**Discussion by apateluk** [vercel/next.js#61773] [Help]
### Summary

I'm using Server Actions in my app and seeking some clarity on how they are exposed and which functions I need to add authentication please ?
If I have two files actions.js and database.js (both in the same folder).
```
action.js
"use server"
import {dbCreateuser} from './database.js'

function checkUsernameisValid(username) {
return username.length > 8;
}

export async function saveUser(user) {
    if (!checkUsernameisValid(user.name)
        return null;

	dbCreateuser(user.name, user.password)
}

export async function saveUser(user) {......
}
```

```
database.js
export async function dbCreateuser(username, password) {
	//code to do database operation on user
}
```
In this very simple example, I have a few questions:
1. Are ALL exported functions from a file marked at the top with "use server" automatically created into end points that can be publicly called ?
2. Anything not marked as export in the same file is not exposed as an end point ?
3. If I have utility functions in another file (in this case database.js) that are exported, are these not created as end points as I have omitted the "use server" ?
My objective is to understand where I need to place an authentication check.
Thanks

**Comment by icyJoseph** [comment]
Hi,

I think you should read the React documentation about `use server`, https://react.dev/reference/react/use-server

Simply put, the `use server` directive, is also an indication to the bundler, that this code, must not be included in JavaScript bundles sent to the client.

Rather, a unique identifier is created, which is used to locate the function in the server. There's no particular endpoint for a given server action, AFAIK, the request is fired against the current route, but with special headers that tell Next.js what function to execute.

> Are ALL exported functions from a file marked at the top with "use server" automatically created into end points that can be publicly called ?

The docs answer to that is:

> Instead of individually marking functions with 'use server', you can add the directive to the top of a file to mark all exports within that file as Server Actions that can be used anywhere, including imported in client code.

So, as long as these are `async` functions, I think they could, potentially be called, that is true. However, to call a server action, you need its unique identifier,  among other things. 

> Anything not marked as export in the same file is not exposed as an end point ?
> If I have utility functions in another file (in this case database.js) that are exported, are these not created as end points as I have omitted the "use server" ?

Here the answer is that it is probably a good idea to collect and export server actions only from dedicated files. Code imported into a `use server` module won't turn into a server action.

And the authentication checks should probably go within the server actions, as first execution, right before you validate the received data.

Please refer to:

- https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#cookies
- https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#security
